rfcs/006-package-management.md

```diff
@@ -0,0 +1,357 @@
+---
+feature: package management
+start-date: 2024-06-28
+author: Joe Neeman
+---
+
+# Package management
+
+People want to reuse code, but Nickel doesn't currently have a good way to do
+it. We should have a way to fetch packages and make them easily available to
+Nickel code. The mechanism needs to be predictable (it should fetch the code
+that the user expects to fetch) and reliable (if it works on my machine it
+should work on your machine).
+
+Explicit non-goals are:
+
+- Management of system (or other non-Nickel) dependencies. You can use Nix for
+  that.
+- Any sort of processing or compilation. Nickel is an interpreted language, so
+  package management is only in charge of distributing unmodified source files.
+- Integration with other language ecosystems; this is only for Nickel code.
+  Anyone integrating Nickel with other languages should use a polyglot build
+  system like Bazel. (However, providing integration points for other tools *is*
+  in scope. More detail on that below.)
+
+## The manifest file
+
+We will require a manifest file in order to import packages. Manifest files
+must be named `electroplate.ncl`, and they are found by searching up from the
+file being evaluated. That is, when the user invokes `nickel export path/to/
+foo.ncl`, we look for a manifest at `path/to/electroplate.ncl` and then at
+`path/electroplate.ncl`, and so on.
+
+The manifest file format is defined by the contract `std.package.Manifest`,
+which is defined as
+
+```nickel
+{
+  name | String,
+  version | Semver,
+  nickel-version | Semver,
```

Comment:
Why is it needed?
Should this be a version constraint? e.g. `>=1`, `^1`, `>1.1.3,<2` etc?

Reply:
Ah, right. I guess `version` should be `SemverConstraint` (or `VersionConstraint`). `nickel-version` is the minimum nickel interpreter version required by the package. Maybe `min-nickel-version` is clearer?